scene/gui/menu_bar.cpp

```diff
@@ -376,7 +376,7 @@ void MenuBar::_notification(int p_what) {
 			}
 
 			Vector2 pos = DisplayServer::get_singleton()->mouse_get_position() - mouse_pos_adjusted - get_global_position();
-			if (pos == old_mouse_pos) {
+			if (pos == Vector2(old_mouse_pos)) {
```

Comment:
Sounds like `old_mouse_pos` should have its type changed instead, this is the only place it's used.

Reply:
@akien-mga Should we make a behavioral change here by changing the type or leave this change as a one for one mechanically and make a second PR with the recommended behavioral change? Either is cool; I'm not sure what is preferred.

Reply:
For an upstream PR it's better to address the source of the issue from the get-go, than to add ad-hoc casts like that. In other words, if we know how to fix the problem, there is no reason to not fix it and do something else, and then make another PR to actually fix it.

Reply:
This includes comments made by bruvzg, btw. It should be preferable to use narrower types than Variant and/or to specify the operator.

Reply:
I'm afraid we may be discussing distinct issues that touch the same lines of code. I'm in no way attempting to fix logic errors or improve performance issues in the context of this change. There are cases where T is being constructed or converted from U, currently implicitly, which cannot be implicitly done in future versions of the language in these contexts. Suggesting that making this behavior explicit somehow doesn't address this problem, but changing which types were constructed would, is to say that both the existing code is wrong by using the incorrect types -- perhaps true, and using the correct types would itself fix the language compatibility problem -- it would not.

Yes, there are numerous instances of questionable behavior caused by implicit conversions especially with regard to the Variant class. They are not limited to comparison operators but are found in assignment, function calls, and wherever the Variant is used in a context that expects a different type. Preventing this behavior is a significant undertaking which requires marking Variant's conversion functions and ctors explicit and then deciding on the behavior you want at each point it was previously being converted from/to and either calling the constructor for the right type or changing the behavior independently at every location. This should itself probably be two steps: one change to make such methods explicit without a behavior change and one to evaluate if the existing conversions are correct. Resolving overload ambiguity in comparison functions is the tip of the iceberg with regards to potential misuse of implicit conversions with the Variant class.

Experience has taught me not to couple behavioral changes with refactoring and I have intentionally not done so here as regressing against such a combined change can be difficult. If there's desire to couple behavior changes with a change that resolves overload ambiguity in future versions of the specification, we can do that, with understanding that it does not resolve the issue that is unintended implicit conversions, doesn't prevent people from doing such things in the future, and that it changes behavior of the existing code. I'd strongly recommend against it though.

Reply:
Strictly speaking, the engine does not benefit from this refactoring, but it would benefit from fixing issues in its code. There is little reason to do a pass that doesn't address the issues, but possibly even hides them better just to preserve some purity and atomic nature of changes. If you are not comfortable adjusting this PR to address the issues highlighted by your research, we can wait for somebody else to do it (but in this case, please open a bug report, so the problem is visible).

> Resolving overload ambiguity in comparison functions is the tip of the iceberg

Let's not get ahead of ourselves. You've made changes that clearly point at specific issues, and you've been advised how to address those issues so that the PR can improve the engine codebase for everyone. There is nothing more to it.